### PR TITLE
[serve] Implement set_options

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

#23157 and #23232 introduce an `Application` object with an immutable deployment dictionary. This change introduces `set_options()`, a deployment method that mutates its attributes. This allows an `Application`'s deployments to be edited after initialization.

## Related issue number

N/A

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
     - New unit tests are added to `test_api.py`.

### CODE REVIEW

File: python/ray/serve/api.py [hunk 1]
```diff
@@ -1257,6 +1260,72 @@ def options(
             _internal=True,
         )
 
+    @PublicAPI
```
Comment (edoakes) [inline python/ray/serve/api.py:1263]:
```suggestion
    @PublicAPI(stability="alpha")
```

**Review by edoakes — APPROVED** [review]
@simon-mo PTAL

File: python/ray/serve/api.py [hunk 2]
```diff
@@ -1257,6 +1260,72 @@ def options(
...
+        self._ray_actor_options = validated._ray_actor_options
+        self._config.autoscaling_config = validated._config.autoscaling_config
+        self._config.graceful_shutdown_wait_loop_s = (
+            validated._config.graceful_shutdown_wait_loop_s
+        )
+        self._config.graceful_shutdown_timeout_s = (
+            validated._config.graceful_shutdown_timeout_s
+        )
+        self._config.health_check_period_s = validated._config.health_check_period_s
+        self._config.health_check_timeout_s = validated._config.health_check_timeout_s
```
Comment (simon-mo) [inline python/ray/serve/api.py:1327]: just `self._config = validated._config`?

Comment (simon-mo) [reply python/ray/serve/api.py:1327]: and the rest of stuff

Comment (shrekris-anyscale) [reply python/ray/serve/api.py]: Good idea– I changed it.

**Review by simon-mo — APPROVED** [review]

**Pull request merged**